pkg/cloud/gcp/actuators/machineset/controller.go

```diff
@@ -0,0 +1,174 @@
+	r.recorder = mgr.GetEventRecorderFor("machineset-controller")
+	r.scheme = mgr.GetScheme()
+	return nil
+}
+
+// Reconcile implements controller runtime Reconciler interface.
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Log.WithValues("machineset", req.Name, "namespace", req.Namespace)
+	logger.V(3).Info("Reconciling")
+
+	ctx := context.Background()
+	machineSet := &machinev1.MachineSet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, machineSet); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Ignore deleted MachineSets, this can happen when foregroundDeletion
+	// is enabled
+	if !machineSet.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+	originalMachineSetToPatch := client.MergeFrom(machineSet.DeepCopy())
+
+	result, err := r.reconcile(machineSet)
+	if err != nil {
+		logger.Error(err, "Failed to reconcile MachineSet")
+		r.recorder.Eventf(machineSet, corev1.EventTypeWarning, "ReconcileError", "%v", err)
+		// we don't return here so we want to attempt to patch the machine regardless of an error.
+	}
+
+	if err := r.Client.Patch(ctx, machineSet, originalMachineSetToPatch); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to patch machineSet: %v", err)
+	}
+
+	if isInvalidConfigurationError(err) {
+		// For situations where requeuing won't help we don't return error.
+		// https://github.com/kubernetes-sigs/controller-runtime/issues/617
+		return result, nil
+	}
+
+	return result, err
+}
+
+func isInvalidConfigurationError(err error) bool {
+	switch t := err.(type) {
+	case *mapierrors.MachineError:
+		if t.Reason == machinev1.InvalidConfigurationMachineError {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *Reconciler) reconcile(machineSet *machinev1.MachineSet) (ctrl.Result, error) {
+	providerConfig, err := getproviderConfig(machineSet)
+	if err != nil {
+		return ctrl.Result{}, mapierrors.InvalidMachineConfiguration("failed to get providerConfig: %v", err)
+	}
+
+	gceService, err := r.getGCPService(machineSet.GetNamespace(), *providerConfig)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	machineType, err := r.cache.getMachineTypeFromCache(gceService, providerConfig.ProjectID, providerConfig.Zone, providerConfig.MachineType)
+	if err != nil {
+		return ctrl.Result{}, mapierrors.InvalidMachineConfiguration("error fetching machine type %q: %v", providerConfig.MachineType, err)
+	}
+
+	if machineSet.Annotations == nil {
+		machineSet.Annotations = make(map[string]string)
+	}
+
+	// TODO: get annotations keys from machine API
+	machineSet.Annotations[cpuKey] = strconv.FormatInt(machineType.GuestCpus, 10)
+	machineSet.Annotations[memoryKey] = strconv.FormatInt(machineType.MemoryMb, 10)
```

Comment:
do we need to add something here for the gpu value?

Reply:
We don't have GPU implemented now so my thought was to leave it empty and allow the autoscaler to do the defaulting when it doesn't find the key.

Then when we implement GPU support we can add the correct value here and everything should carry on working.

I don't think we should be setting it to 0 explicitly here in case users manually add them, until we have a way to support, them omitting makes more sense to me

Reply:
sounds good to me!

> I don't think we should be setting it to 0 explicitly here in case users manually add them, until we have a way to support, them omitting makes more sense to me

agreed